learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,168 @@
+import recordclass
+import string
+import aiohttp
+import aiohttp.web
+import learning_observer
+import learning_observer.runtime
+
+
+# These took a while to find, but many are documented here:
+# https://developers.google.com/drive/api/v3/reference/
+# This list might change. Many of these contain additional (optional) parameters
+# which we might add later. This is here for debugging, mostly. We'll stabilize
+# APIs later.
+class Endpoint(recordclass.make_dataclass("Endpoint", ["name", "remote_url", "doc", "cleaners", "lms"], defaults=["", None])):
+    def arguments(self):
+        return extract_parameters_from_format_string(self.remote_url)
+
+    def _local_url(self):
+        parameters = "}/{".join(self.arguments())
+        base_url = f"/{self.lms}/{self.name}"
+        if len(parameters) == 0:
+            return base_url
+        else:
+            return base_url + "/{" + parameters + "}"
+
+    def _add_cleaner(self, name, cleaner):
+        if self.cleaners is None:
+            self.cleaners = dict()
+        self.cleaners[name] = cleaner
+        if 'local_url' not in cleaner:
+            cleaner['local_url'] = self._local_url + "/" + name
+
+    def _cleaners(self):
+        if self.cleaners is None:
+            return []
+        else:
+            return self.cleaners
+
+def extract_parameters_from_format_string(format_string):
+    '''
+    Extracts parameters from a format string. E.g.
+    >>> ("hello {hi} my {bye}")]
+    ['hi', 'bye']
+    '''
+    return [f[1] for f in string.Formatter().parse(format_string) if f[1] is not None]
+
+def raw_access_partial(raw_ajax, remote_url, name=None):
+    '''
+    This is a helper which allows us to create a function which calls specific
+    Google APIs.
+
+    To test this, try:
+
+        print(await raw_document(request, documentId="some_google_doc_id"))
+    '''
+    async def caller(request, **kwargs):
+        '''
+        Make an AJAX request to Google
+        '''
+        return await raw_ajax(request, remote_url, **kwargs)
+    setattr(caller, "__qualname__", name)
+
+    return caller
+
+def api_docs_handler(request, ENDPOINTS):
+    '''
+    Return a list of available endpoints.
+
+    Eventually, we should also document available function calls
+    '''
+    response = "URL Endpoints:\n\n"
+    for endpoint in ENDPOINTS:
+        response += f"{endpoint._local_url()}\n"
+        cleaners = endpoint._cleaners()
+        for c in cleaners:
+            response += f"   {cleaners[c]['local_url']}\n"
+    response += "\n\n Globals:"
+    return aiohttp.web.Response(text=response)
+
+def register_cleaner(data_source, cleaner_name, ENDPOINTS):
```

Comment:
Fully capitalized variables are used for constants. They should not be used for parameter names.
Note that parameters might default to a constant.

Reply:
Fixed